File: src/main/java/com/mpe85/grampa/rule/Command.java
package com.mpe85.grampa.rule;

/**
 * A parser command. A command is a special {@link Action} that always succeeds (i.e. always returns true)
 * 
 * @author mpe85
 *
 * @param <T>
 *        the type of the stack elements
 */
@FunctionalInterface
public interface Command<T> {
	
	/**
	 * Executes the parser command.
	 * 
	 * @param context
	 *                an action context
	 */
	void execute(ActionContext<T> context);
	
	/**
	 * Converts the parser command to a parser action.
	 * 
	 * @return a parser command
	 */
	public default Action<T> toAction() {
		return ctx -> {
			this.execute(ctx);
			return true;
		};
	}
	
}

File: src/main/kotlin/com/mpe85/grampa/rule/Command.kt
package com.mpe85.grampa.rule

import com.mpe85.grampa.context.RuleContext

/**
 * A parser command. A command is a special [Action] that always succeeds (i.e. always returns true).
 *
 * @author mpe85
 * @param[T] The type of the stack elements
 */
@FunctionalInterface
public fun interface Command<T> {

    /**
     * Execute the parser command.
     *
     * @param[context] A rule context
     */
    public fun execute(context: RuleContext<T>)
}

/**
 * Convert the parser command to a parser action.
 *
 * @return A parser action
 */
public fun <T> Command<T>.toAction(): Action<T> = Action<T> { execute(it); true }
